src/runtime/objmodel.cpp

```diff
@@ -5446,6 +5420,100 @@ Box* typeNew(Box* _cls, Box* arg1, Box* arg2, Box** _args) {
     return made;
 }
 
+// Analogous to CPython's type_new.
+// This is assigned directly to type_cls's (PyType_Type's) tp_new slot and skips
+// doing an attribute lookup for __new__.
+//
+// We need this to support some edge cases. For example, in ctypes, we have a function:
+// PyCSimpleType_new(PyTypeObject *type, PyObject *args, PyObject *kwds)
+// {
+//     ...
+//     result = (PyTypeObject *)PyType_Type.tp_new(type, args, kwds);
+//     ...
+// }
+//
+// Assigned to the tp_new of PyCSimpleType, a metaclass. By calling PyType_Type.tp_new,
+// we don't want to do an attribute lookup for __new__, because that would end up calling
+// the tp_new of the subclass PyCSimpleType (PyCSimpleType_new again) and end up in a loop.
+Box* typeNewType(BoxedClass* metatype, Box* args, Box* kwds) noexcept {
```

Comment:
I think this should be called `type_new` (ie using the cpython convention for cpython-like functions)